File: 25/solution.py
import networkx as nx

with open("input.txt", "r") as f:
  lines = f.readlines()

G = nx.Graph()
nodes = set()

for line in lines:
  frNode, line = line.split(":")
  toNodes = line.split()

  nodes.add(frNode)
  G.add_node(frNode)

  for toNode in toNodes:
    nodes.add(toNode)
    G.add_node(toNode)
    G.add_edge(frNode, toNode, capacity=1)

nodes = list(nodes)
cut, part = nx.minimum_cut(G, nodes[0], nodes[3])

print(cut)
print(len(part[0]) * len(part[1]))

File: 25/solution.go
package puzzle25

import (
	"advent2023/internal/puzzle"
	"advent2023/util"
	"strings"

	"github.com/dominikbraun/graph"
)

type p struct {
	// nodes []string
	g graph.Graph[string, string]
}

func (p *p) Init(filename string) {
	lines := util.GetLines(filename)
	nodes := map[string]bool{}
	p.g = graph.New(graph.StringHash)
	// p.nodeIdx = make(map[string]int)
	// p.adj = make(map[int]map[int]bool)

	for _, line := range lines {
		sp := strings.Split(line, ":")
		from := strings.TrimSpace(sp[0])
		to := strings.Fields(sp[1])
		if !nodes[from] {
			nodes[from] = true
			p.g.AddVertex(from)
		}
		for _, t := range to {
			if !nodes[t] {
				nodes[t] = true
				p.g.AddVertex(t)
			}
			p.g.AddEdge(from, t)
		}
	}
}

func (p *p) Solve1() any {
	return "DONT WORK YET. LIFE IS SHORT, JUST USE PYTHON."
}

func (p *p) Solve2() any {
	return "THANK YOU MR.WASTL!"
}

func init() {
	puzzle.Register(25, &p{})
}
